### ISSUE TITLE
[unused_declaration rule] `// swiftlint:enable unused_declaration` does not work

### ISSUE THREAD

**Issue by acecilia** [realm/SwiftLint#3769]
### New Issue Checklist

- [x] Updated SwiftLint to the latest version
- [x] I searched for [existing GitHub issues](https://github.com/realm/SwiftLint/issues)

### Describe the bug

When running the `unused_declaration` analyze rule over below code, there is no violation being reported:

```swift
// swiftlint:disable all
// swiftlint:enable unused_declaration
private func unusedPrivate() { }
```

### Environment

* SwiftLint version (run `swiftlint version` to be sure)? `0.45.0`
* Installation method used (Homebrew, CocoaPods, building from source, etc)? `mint`
* Paste your configuration file:

```yml
analyzer_rules:
  - unused_declaration
```

* Are you using [nested configurations](https://github.com/realm/SwiftLint#nested-configurations)?
  If so, paste their relative paths and respective contents. `No`
* Which Xcode version are you using (check `xcodebuild -version`)? `Xcode 12.5.1`
* Do you have a sample that shows the issue? Yes. Clone https://github.com/acecilia/swift-compilation-database and then do `make go_with_xcodebuild`. It is possible to observe that [here](https://github.com/acecilia/swift-compilation-database/blob/d9b237561e2d9f1232ac852e51ad762405b8548b/Sources/ModuleA/ModuleA1.swift#L12) there is no violation reported

**Comment by mildm8nnered** [comment]
I think this is actually working as intended.

If I remove `// swiftlint:disable all`, the `unused_declaration` triggers for me.

Possible the confusion stems from the `// swiftlint:enable unused_declaration` - I believe SwiftLint will ignore this because of the `// swiftlint:disable all`. This will disable all rules until a matching `// swiftlint:enable all` is seen - rules can't be re-enabled one-by-one.

A couple of interesting oddities I came across while playing with this:

In the following code:

```
import Cocoa

// swiftlint:disable all
// swiftlint:enable unused_declaration
private func unusedPrivate() { }

@main
class AppDelegate: NSObject, NSApplicationDelegate {
    @IBOutlet var window: NSWindow!
    
    // swiftlint:disable:next nonexistent_rule
    let y = 0
    
    // swiftlint:disable:next force_try
    let z = 0
    
    // swiftlint:enable:next force_try
}
```

Even though we have `// swiftlint:disable all` on, `superfluous_disable_command` will get triggered by the `// swiftlint:disable:next nonexistent_rule`, referencing a rule that doesn't exist.

Also, if I change `// swiftlint:disable all` to, for example, `// shjkljklklswiftlint:disable all`, that will also disable all rules, which seems a bit off.

**Comment by SimplyDanny** [comment]
@mildm8nnered: Could you create separate tickets for the oddities you discovered if they aren't addressed somewhere else already?

Apart from that, you are right with:

> Possible the confusion stems from the `// swiftlint:enable unused_declaration` - I believe SwiftLint will ignore this because of the `// swiftlint:disable all`. This will disable all rules until a matching `// swiftlint:enable all` is seen - rules can't be re-enabled one-by-one.

**Comment by acecilia** [comment]
> rules can't be re-enabled one-by-one

Understood 🙌 . Then, how could I disable all rules for linting, but still be able to detect unused declarations? 

Usecase: autogenerated code, where the linting is not important, but the analysis of unused code is

**Comment by mildm8nnered** [comment]
> > rules can't be re-enabled one-by-one
> 
> Understood 🙌 . Then, how could I disable all rules for linting, but still be able to detect unused declarations?
> 
> Usecase: autogenerated code, where the linting is not important, but the analysis of unused code is

So `unused_declaration` is an analyzer rule - in an analyzer pass, only the analyzer rules, of which there are only a few, will be applied.

```
+------------------------------------------+--------+-------------+------------------------+-------------+----------+----------------+---------------+
| identifier                               | opt-in | correctable | enabled in your config | kind        | analyzer | uses sourcekit | configuration |
+------------------------------------------+--------+-------------+------------------------+-------------+----------+----------------+---------------+
| capture_variable                         | yes    | no          | no                     | lint        | yes      | yes            | warning       |
| explicit_self                            | yes    | yes         | no                     | style       | yes      | yes            | warning       |
| typesafe_array_init                      | yes    | no          | no                     | lint        | yes      | yes            | warning       |
| unused_declaration                       | yes    | no          | no                     | lint        | yes      | yes            | severity: ... |
| unused_import                            | yes    | yes         | no                     | lint        | yes      | yes            | severity: ... |
```

So typically, you would use linting normally, and then less frequently most likely, run the analyzer, which is much much slower (seconds for linting, vs. 10s of minutes for analysing, on an X hundred thousand line codebase).

A little bit confusingly, you can enable the analyzer rules in the `opt_in_rules` section, or in their own `analyzer_rules` section in the configuration file.

But, you don't need to worry about the ordinary linting rules - they won't be applied, so you don't need to disable them.

See https://github.com/realm/SwiftLint#analyze for more info.

**Comment by acecilia** [comment]
> But, you don't need to worry about the ordinary linting rules - they won't be applied, so you don't need to disable them.

It is very common for autogenerated code to start by having `// swiftlint:disable all` at the top of the file, to disable linting it.

If you run the analyzer in such file, it wont detect unused declarations, which is my main issue

**Comment by SimplyDanny** [comment]
I get your point. Unfortunately, there is currently no way for you to achieve this at the moment.

On the other hand, generated code could equally follow the style and quality claims of hand-written code. Developers also interacted and read it. So why disable linting such files?

**Comment by acecilia** [comment]
@SimplyDanny As a consumer of generated code, in many cases you do not have control of the generated code, neither its formatting. See for example the templates used by Sourcery or SwiftGen

**Comment by mildm8nnered** [comment]
@acecilia - I'm pretty much in the same situation as you - generated code that can't easily be linted.

We exclude those directories from linting (which means we don't need to `disable all` at the top).

But as analyzer runs are separate, you could include those directories in analyzer runs, if you were only interested in `unused_declaration`.
